src/__test__/extra/typedEvents-more.test.ts

```diff
@@ -0,0 +1,296 @@
+      productItems: ProductItems;
+      openedAt: Date;
+      confirmedAt?: Date;
+    }
+
+    const enum ShoppingCartErrors {
+      OPENED_EXISTING_CART = "OPENED_EXISTING_CART",
+      CART_NOT_FOUND = "CART_NOT_FOUND",
+      PRODUCT_ITEM_NOT_FOUND = "PRODUCT_ITEM_NOT_FOUND",
+      UNKNOWN_EVENT_TYPE = "UNKNOWN_EVENT_TYPE",
+    }
+
+    const addProductItem = (
+      inventory: ProductItems,
+      { productId, quantity }: ProductItem
+    ): ProductItems => {
+      const current = inventory.get(productId);
+
+      if (!current) {
+        return inventory.set(productId, { productId, quantity });
+      }
+
+      return inventory.set(productId, {
+        ...current,
+        quantity: current.quantity + quantity,
+      });
+    };
+
+    const removeProductItem = (
+      inventory: ProductItems,
+      { productId, quantity }: ProductItem
+    ): ProductItems => {
+      const current = inventory.get(productId);
+
+      if (!current || current.quantity < quantity) {
+        throw ShoppingCartErrors.PRODUCT_ITEM_NOT_FOUND;
+      }
+
+      if (current.quantity === quantity) {
+        inventory.delete(productId);
+        return inventory;
+      }
+
+      return inventory.set(productId, {
+        ...current,
+        quantity: current.quantity - quantity,
+      });
+    };
+
+    const shoppingCartAggrigator = createStreamAggrigator<
+      ShoppingCart,
+      ShoppingCartEvent
+    >((currentState, event) => {
+      if (event.type === "shopping-cart-opened") {
+        if (currentState != null) throw ShoppingCartErrors.OPENED_EXISTING_CART;
+        return {
+          id: event.data.shoppingCartId,
+          clientId: event.data.clientId,
+          openedAt: new Date(event.data.openedAt),
+          productItems: new Map(),
+          status: ShoppingCartStatus.Opened,
+        };
+      }
+
+      if (currentState == null) throw ShoppingCartErrors.CART_NOT_FOUND;
+
+      switch (event.type) {
+        case "product-item-added-to-shopping-cart":
+          return {
+            ...currentState,
+            productItems: addProductItem(
+              currentState.productItems,
+              event.data.productItem
+            ),
+          };
+        case "product-item-removed-from-shopping-cart":
+          return {
+            ...currentState,
+            productItems: removeProductItem(
+              currentState.productItems,
+              event.data.productItem
+            ),
+          };
+        case "shopping-cart-confirmed":
+          return {
+            ...currentState,
+            status: ShoppingCartStatus.Confirmed,
+            confirmedAt: new Date(event.data.confirmedAt),
+          };
+        default: {
+          const _: never = event;
+          throw ShoppingCartErrors.UNKNOWN_EVENT_TYPE;
+        }
+      }
+    });
+
+    const enum ProductsIds {
+      T_SHIRT = "team-building-excercise-2022",
+      SHOES = "air-jordan",
+    }
+
+    const clientId = "client-123";
```

Comment:
**SUGGESTION:** Better to not use dashes in ids if it's not needed, to not accidentally conflict with the stream name category split.
```suggestion
    const clientId = "client_123";
```